include/seastar/util/sampler.hh

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This file has been originally been imported from:
+// https://cs.android.com/android/platform/superproject/+/013901367630d3ec71c9f2bb3f3077bd11585301:external/perfetto/src/profiling/memory/sampler.h
+
+//
+// The code has been modified as follows:
+//
+//  - Integrated into seastar and adapted to coding style
+//  - Right now we don't account for samples multiple times (in case we have
+//  multiple loops of drawing from the exp distribution). The reason is that
+//  in our memory sampler we would have to store the weight in addition to the
+//  alloation site ptr as on free we need to know how much a sample accounted
+//  for. Hence, for now we simply always use the sampling interval.
```

Comment:
So we can compensate for this in post-processing, right? Like we know the expected number of times an allocation would be sampled (contingent on it being sampled once) so we can "gross up" the weight at that point.

Reply:
Yes, in post-processing or as it's currently done in the size tracking directly.